### ISSUE TITLE
Dev / ui back testing - Javascript heap out of memory

### ISSUE THREAD

**Issue by thegamecat** [askmike/gekko#603]
Testing the back test on the dev branch, hope this helps. 

Meta:
The sqlite db is about 5 years and 10 months of data
Running standard StochRSI
Candle size 1 minute
History size 10 (not sure what this is for.

This is on windows 10, machine has 16gig ram and memory allocation in windows is 20gig.

Returns the following error in the console:

- <--- Last few GCs --->

  617021 ms: Mark-sweep 1383.9 (1435.2) -> 1383.8 (1435.2) MB, 2113.1 / 0.0 ms [allocation failure] [scavenge might not succeed].
  619173 ms: Mark-sweep 1383.8 (1435.2) -> 1383.8 (1435.2) MB, 2151.9 / 0.0 ms [allocation failure] [scavenge might not succeed].
  621241 ms: Mark-sweep 1383.8 (1435.2) -> 1385.1 (1419.2) MB, 2067.9 / 0.0 ms [last resort gc].
  623359 ms: Mark-sweep 1385.1 (1419.2) -> 1386.4 (1419.2) MB, 2117.6 / 0.0 ms [last resort gc].

<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 00000143079CFB61 <JS Object>
    1: stringify [native json.js:178] [pc=000003C553F384F7] (this=00000143079C9121 <a JSON with map 000002AEB190A0E1>,E=00000225FF6B6581 <an Object with map 000002AEB1910CD1>,F=0000014307904381 <undefined>,S=0000014307904381 <undefined>)
    2: arguments adaptor frame: 1->3
    3: get length [C:\Bots\gekko-develop\node_modules\koa\lib\response.js:204] [pc=000003C554213886] (this=000002DCE9440...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory

**Comment by askmike** [comment]
Where you running a backtest or a live gekko?

**Comment by thegamecat** [comment]
Backtest.

**Comment by askmike** [comment]
Wow! How big is your database (the history folder within gekko)?

**Comment by thegamecat** [comment]
450 meg. It's actually not that bad. The awesome thing is your back tester works like a charm, although I am thinking about how I can validate that it is working properly. As your message says, the backtester is in testing so....

**Comment by thegamecat** [comment]
I'm not getting errors any more but the the UI essentially stops after while, both with the large history and a shorter one of eth for a year.

The process does continue though, but doesn't report back when it completes.

**Comment by askmike** [comment]
If you set the candleSize to one day, does that still happen? Also can you take a screenshot of where it gets stuck?

**Comment by thegamecat** [comment]
Example: 
poloniex	BTC	XEM	2016-09-11 21:59	2017-03-11 20:59	5 months, 4 weeks, 18 hours, 30 minutes

Basically nothing happens visually aside from the Backtest button appears. Nothing different appears in Node. 

If I set it to 1 day it works fine. 

Looks like a timeout somewhere on the web client, this works fine on CLI.

**Comment by askmike** [comment]
The lower you set the candle size, the more processing Gekko has to do. But my pretty old  laptop (macbook air from 2012) has no problem doing a backtest with the UI using over 8 months of data & 5 minute candles.

------

The webclient does a backtest "request" to a "gekko server" that will actually perform the backtest. this request (a HTTP POST request) is where the timeout most likely happened, there are no timeouts in the web UI.

Questions:

- What browser are you using?
- Could you give a rough estimate of your PC/laptop?
- What OS are you using?

If you are using chrome, can you do the following:

- load the UI
- right click on the page - click "inspect"
- go to the "network" tab
- start a backtest
- wait a few minutes
- make a screenshot of the result

**Comment by thegamecat** [comment]
<img width="1115" alt="screen shot 2017-03-12 at 16 26 41" src="https://cloud.githubusercontent.com/assets/3260602/23833684/e9df64d8-0740-11e7-97bf-b3b8a9bba517.png">

MBP 2015 - OSC (latest) - 8 or 16gig memory, forget which and an i5.

To clarify, that is the screen after the backtest has been run / fails.

**Comment by thegamecat** [comment]
I'm happy to screenshare / skype whatever if you like.

**Comment by askmike** [comment]
I thought I would be able to fix this fast, but it appears I am not..

So I am able to backtest MACD against 8 months of data with 1 minute candle size, but the UI will freeze for me. Interesting that your laptop, which should be at least twice as fast as mine, struggles this hard (I have 4gb of RAM, which I think is my problem).

So this means that Gekko is quite able to backtest over so much data, but it's hard to get the results into the UI (all data needed to spawn the price chart, which in your case is ~6 months x 30 days x 24 hours x 60 minutes = 259,200 data points).

-----

I will look into this after #611, which has my main priority now. But note that I am not sure what you are hoping to achieve by running a strat against 1 minute candles: I have never seen someone making a successful TA strategy that operates on the 1 minute level, there is to much micro market noise. And even if you want to do that, there is really no big reason to simulate over 6 months of data (in my opinion). So I guess this might be out of scope for what Gekko is intended to do.

In my experience most successful strats run between 30 minute and 3 day candles.

**Comment by thegamecat** [comment]
I know what you mean but I have a strat which works great on 1m candles.

It also fails on a windows machine (as per the original post).

I've also had the same issue with 5m candles. 

I've done some thinking about how to work around this based on the same 6 month logic as you have - if I could select the date range then this would likely solve this.

Out of interest, which db are you using? I'm using sqlite.

And finally, understand the priority lies elsewhere.

**Comment by askmike** [comment]
> It also fails on a windows machine (as per the original post).

Even though the error is different, the solution is the same: if it's a huge backtest we need to send back less detailed graph, so that the UI can handle it (a graph with over 100,000 points is not very readable anyway).

> I've done some thinking about how to work around this based on the same 6 month logic as you have - if I could select the date range then this would likely solve this.

That used to be an option before, I will add a daterange picker back to the backtester!

> Out of interest, which db are you using? I'm using sqlite.

Yes sqlite as well, I am not sure if the other ones support all the new requirements (they were added by other people, before I started working on the UI).

**Comment by thegamecat** [comment]
Thanks. That date range would be brilliant.

What I'm doing at the moment is writing all outputs in a method to a csv - covering several different indicators. Then I'm segmenting that manually and making my own graphs with it in Tableau. This is fundamentally because I am a trading noob and this, combined with coding the methods helps me see what's working and what isn't. That's why I'm looking at 1m plus long periods of time. 

I can also use this output for some machine learning when I get to it. 

It's a bloody great project and tool you've made :)

**Comment by askmike** [comment]
Closing this for now since we can't reproduce under normal usage, there are some performance optimalisations on the roadmap :)
